### PR TITLE
[#11706] fix(catalog-hive): replace switch-on-enum with if-else in HiveCatalogCapability.caseSensitiveOnName()

### LINKED ISSUE #11706: [Bug report] bug(catalog-hive): HiveCatalogCapability.caseSensitiveOnName() uses switch-on-enum which generates a synthetic $1 class, causing NoClassDefFoundError when called outside IsolatedClassLoader context
### Version

main branch

### Bug Description

`HiveCatalogCapability.caseSensitiveOnName()` uses a `switch` statement on
`Capability.Scope` (an enum). The Java compiler generates a synthetic helper
class `HiveCatalogCapability$1` for every `switch`-on-enum to maintain a
stable ordinal mapping table.

`caseSensitiveOnName()` is called by `SchemaNormalizeDispatcher` **outside**
the `IsolatedClassLoader.withClassLoader()` boundary:

```
SchemaNormalizeDispatcher.normalizeCaseSensitive()
  → getCapability() → c.capabilities()
      → withClassLoader { catalog.capability() }  ← IsolatedClassLoader context; creates HiveCatalogCapability instance
    ← withClassLoader exits; thread context classloader restored to server classloader
  → applyCaseSensitive() → capabilities.caseSensitiveOnName(Scope.SCHEMA)
      → JVM attempts to load HiveCatalogCapability$1
```

When `HiveCatalogCapability` is correctly loaded by `IsolatedClassLoader`,
its defining classloader is `IsolatedClassLoader`, and `$1` is loaded from
`execJars` successfully. However, in certain classloader initialization timing
windows (e.g. concurrent requests during catalog cache rebuild), `$1` is
requested while the server classloader is involved in the loading chain.
**The JVM permanently caches this load failure**, and all subsequent calls
throw `NoClassDefFoundError` until the process is restarted.

### Root Cause

`SchemaNormalizeDispatcher` calls `Capability.caseSensitiveOnName()` — a
method on a catalog-side implementation — **after** returning from
`withClassLoader`. This is an architectural inconsistency: catalog
implementation code is invoked outside the catalog's isolated classloader
context.

The immediate trigger is the compiler-generated `$1` synthetic class, which
only exists because `switch`-on-enum is used. Replacing `switch` with
`if`-`else` (using `==` on enum constants) eliminates `$1` entirely and makes
`caseSensitiveOnName()` safe to call from any classloader context.

### Proposed Fix

Replace `switch` with `if`-`else` in `HiveCatalogCapability.caseSensitiveOnName()`:

```java
// Before
switch (scope) {
  case SCHEMA: case TABLE: case COLUMN:
    return CapabilityResult.unsupported("Hive is case insensitive.");
  default:
    return CapabilityResult.SUPPORTED;
}

// After
if (scope == Scope.SCHEMA || scope == Scope.TABLE || scope == Scope.COLUMN) {
  return CapabilityResult.unsupported("Hive is case insensitive.");
}
return CapabilityResult.SUPPORTED;
```

This removes the `$1` synthetic class entirely. `==` on enum constants
compiles to `if_acmpeq` (reference comparison), which has no synthetic class
dependency and is safe across classloader boundaries.

### Additional Context

The same pattern should be reviewed in other `Capability` implementations
across catalogs, as any `switch`-on-enum in a catalog-side class called from
`SchemaNormalizeDispatcher` / `TableNormalizeDispatcher` (outside
`withClassLoader`) would be susceptible to the same issue.

### Error message and/or stacktrace

java.lang.NoClassDefFoundError: org/apache/gravitino/catalog/hive/HiveCatalogCapability$1
        at org.apache.gravitino.catalog.hive.HiveCatalogCapability.caseSensitiveOnName(HiveCatalogCapability.java:45)
        at org.apache.gravitino.catalog.CapabilityHelpers.applyCaseSensitiveOnName(CapabilityHelpers.java:476)
        at org.apache.gravitino.catalog.CapabilityHelpers.applyCaseSensitive(CapabilityHelpers.java:123)
        at org.apache.gravitino.catalog.SchemaNormalizeDispatcher.normalizeCaseSensitive(SchemaNormalizeDispatcher.java:94)
        at org.apache.gravitino.catalog.SchemaNormalizeDispatcher.loadSchema(SchemaNormalizeDispatcher.java:71)
        at org.apache.gravitino.listener.SchemaEventDispatcher.loadSchema(SchemaEventDispatcher.java:116)
        ...
        at org.apache.gravitino.utils.PrincipalUtils.doAs(PrincipalUtils.java:44)

### How to reproduce

+ Which Gravitino version to use
main / any version with Hive catalog

+ Steps
1. Deploy Gravitino with a Hive catalog configured
2. Call any API that goes through SchemaNormalizeDispatcher
   (e.g. loadSchema, schemaExists, loadTable on first access):

   GET /api/metalakes/{metalake}/catalogs/{hive-catalog}/schemas/{schema}

3. Under specific JVM classloader initialization timing, observe:
   NoClassDefFoundError: org/apache/gravitino/catalog/hive/HiveCatalogCapability$1

4. All subsequent requests to the same catalog on this process instance
   fail with the same error until restart.

+ Root cause
SchemaNormalizeDispatcher calls capabilities.caseSensitiveOnName() AFTER
IsolatedClassLoader.withClassLoader() exits (i.e. outside the catalog's
isolated classloader context). The switch-on-enum in caseSensitiveOnName()
causes javac to generate HiveCatalogCapability$1. In certain classloader
initialization timing windows the server classloader is involved in the
loading chain and cannot find $1; the JVM permanently caches this failure.

+ Note
The bug is intermittent because it depends on classloader initialization
timing. Once triggered it is permanent within the process lifetime.
Restarting the process clears the cached failure.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JoegenUSTC**
### What changes were proposed in this pull request?

Replace the `switch`-on-enum with `if`-`else` in
`HiveCatalogCapability.caseSensitiveOnName()`, and add unit tests to
verify both the functional behavior and the absence of the synthetic
`$1` class in the compiled output.

### Why are the changes needed?

Fix: #11706

`SchemaNormalizeDispatcher` calls `caseSensitiveOnName()` **after**
`IsolatedClassLoader.withClassLoader()` exits — i.e. outside the catalog's
isolated classloader context. The `switch`-on-enum causes the Java compiler
to generate a synthetic helper class `HiveCatalogCapability$1`. In certain
classloader initialization timing windows, the server classloader is involved
in loading `$1` but cannot find it in the server classpath. **The JVM
permanently caches this load failure for the process lifetime**, causing all
subsequent calls to throw `NoClassDefFoundError` until restart.

Replacing `switch` with `if`-`else` using `==` on enum constants compiles to
`if_acmpeq` (reference comparison) with no synthetic class dependency, making
`caseSensitiveOnName()` safe to call from any classloader context.

### Does this PR introduce _any_ user-facing change?

No. The method behavior is identical; only the compiled representation changes.

### How was this patch tested?

Added `TestHiveCatalogCapability` with three test methods:
- Case-insensitive scopes (`SCHEMA`, `TABLE`, `COLUMN`) return `unsupported`
- Case-sensitive scopes (`FILESET`, `TOPIC`, `PARTITION`, `MODEL`) return `supported`
- `HiveCatalogCapability$1` does not exist in the compiled output
  (guards against future regression back to `switch`)

### CODE REVIEW

File: catalogs/catalog-hive/src/main/java/org/apache/gravitino/catalog/hive/HiveCatalogCapability.java [hunk 1]
```diff
@@ -42,15 +42,16 @@ public CapabilityResult columnDefaultValue() {
...
+    // SchemaNormalizeDispatcher calls this method outside the IsolatedClassLoader.withClassLoader()
+    // boundary. A switch-on-enum causes the JVM to load $1 via the server classloader, which
+    // cannot find it, and the JVM permanently caches the failure for the process lifetime.
+    // if-else with == compiles to if_acmpeq (reference comparison) with no synthetic class.
+    // Hive is case insensitive, see
+    // https://cwiki.apache.org/confluence/display/Hive/User+FAQ#UserFAQ-AreHiveSQLidentifiers(e.g.tablenames,columnnames,etc)casesensitive?
+    if (scope == Scope.SCHEMA || scope == Scope.TABLE || scope == Scope.COLUMN) {
+      return CapabilityResult.unsupported("Hive is case insensitive.");
     }
+    return CapabilityResult.SUPPORTED;
```
[the remark this answers is not in the document]
Comment (JoegenUSTC) [reply catalogs/catalog-hive/src/main/java/org/apache/gravitino/catalog/hive/HiveCatalogCapability.java:57]: Good catch! You're right that the behavior change is meaningful — the original switch would throw NullPointerException for a null scope (per JLS), while the if-else would silently fall through and return SUPPORTED, which is semantically wrong. I've added an explicit throw new NullPointerException("scope") guard to preserve the original fail-fast behavior.

File: catalogs/catalog-hive/src/test/java/org/apache/gravitino/catalog/hive/TestHiveCatalogCapability.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  public void testCaseSensitiveOnNameForCaseSensitiveScopes() {
+    // All other scopes are case sensitive (default behaviour).
+    Assertions.assertTrue(capability.caseSensitiveOnName(Capability.Scope.FILESET).supported());
+    Assertions.assertTrue(capability.caseSensitiveOnName(Capability.Scope.TOPIC).supported());
+    Assertions.assertTrue(capability.caseSensitiveOnName(Capability.Scope.PARTITION).supported());
+    Assertions.assertTrue(capability.caseSensitiveOnName(Capability.Scope.MODEL).supported());
+  }
+
+  @Test
+  public void testCaseSensitiveOnNameSyntheticClassAbsent() throws ClassNotFoundException {
```
[the remark this answers is not in the document]
Comment (JoegenUSTC) [reply catalogs/catalog-hive/src/test/java/org/apache/gravitino/catalog/hive/TestHiveCatalogCapability.java]: Agreed, the throws ClassNotFoundException declaration is redundant here since assertThrows fully handles the exception and it can never escape the test method. Removed.

File: catalogs/catalog-hive/src/main/java/org/apache/gravitino/catalog/hive/HiveCatalogCapability.java [hunk 3]
```diff
@@ -42,15 +42,19 @@ public CapabilityResult columnDefaultValue() {
...
+    // Use if-else instead of switch-on-enum to avoid the compiler-generated synthetic class $1.
+    // SchemaNormalizeDispatcher calls this method outside the IsolatedClassLoader.withClassLoader()
+    // boundary. A switch-on-enum causes the JVM to load $1 via the server classloader, which
+    // cannot find it, and the JVM permanently caches the failure for the process lifetime.
+    // if-else with == compiles to if_acmpeq (reference comparison) with no synthetic class.
+    // Hive is case insensitive, see
+    // https://cwiki.apache.org/confluence/display/Hive/User+FAQ#UserFAQ-AreHiveSQLidentifiers(e.g.tablenames,columnnames,etc)casesensitive?
+    if (scope == null) {
+      throw new NullPointerException("scope");
     }
```
Comment (diqiu50) [inline catalogs/catalog-hive/src/main/java/org/apache/gravitino/catalog/hive/HiveCatalogCapability.java]: Scope should not be null,  please using `Preconditions.checkArgument`

**Review by diqiu50 — COMMENTED** [review]
Thanks for your contribution The fix is correct, but it's a workaround rather than a root fix. The real issue is that CatalogWrapper.capabilities() returns a catalog-classloader-loaded object that is then used outside withClassLoader() — any method on a Capability implementation that touches plugin-jar classes will hit the same problem. Worth a follow-up issue to ensure all Capability method calls are made within the correct classloader context.

**Comment by yuqi1129** [comment]
If this problem did exist for the Hive catalog, we may need to check all catalogs to avoid similar problem.

File: catalogs/catalog-hive/src/main/java/org/apache/gravitino/catalog/hive/HiveCatalogCapability.java (hunk 3 above)
Comment (JoegenUSTC) [reply catalogs/catalog-hive/src/main/java/org/apache/gravitino/catalog/hive/HiveCatalogCapability.java]: Good point, updated to use Preconditions.checkArgument(scope != null, "scope cannot be null") as per the Gravitino coding convention. Also added a dedicated test testCaseSensitiveOnNameNullScopeThrows to cover the null scope case.

**Comment by JoegenUSTC** [comment]
> same

Thanks for the thorough review! Agreed — this is a workaround rather than a root fix. I've opened a follow-up issue #11726 to track ensuring all Capability method calls are made within the correct classloader context.
@diqiu50

**Comment by JoegenUSTC** [comment]
> If this problem did exist for the Hive catalog, we may need to check all catalogs to avoid similar problem.

Thanks for the note! That's a valid concern — any catalog Capability implementation that references plugin-jar classes could hit the same issue. I've opened #11726 to track a proper root fix. Looking forward to your investigation, and happy to help if needed.

**Comment by JoegenUSTC** [comment]
Hi @diqiu50 @yuqi1129, thanks for the valuable feedback!

Two updates:
1. **Null check**: Updated to use `Preconditions.checkArgument(scope != null, "scope cannot be null")` per the Gravitino coding convention, with a dedicated test added to cover the null scope case.
2. **Root fix**: Opened follow-up issue #11726 to track ensuring all `Capability` method calls are made within the correct classloader context — agreed this PR is a workaround and the broader fix deserves its own issue.

CI is now passing. Could this PR be merged as-is, with the root fix tracked in #11726?

**Comment by yuqi1129** [comment]
I have no more comments.

**Comment by diqiu50** [comment]
> Hi @diqiu50 @yuqi1129, thanks for the valuable feedback!
> 
> Two updates:
> 
> 1. **Null check**: Updated to use `Preconditions.checkArgument(scope != null, "scope cannot be null")` per the Gravitino coding convention, with a dedicated test added to cover the null scope case.
> 2. **Root fix**: Opened follow-up issue [[Improvement] Ensure all Capability method calls are made within the correct classloader context #11726](https://github.com/apache/gravitino/issues/11726) to track ensuring all `Capability` method calls are made within the correct classloader context — agreed this PR is a workaround and the broader fix deserves its own issue.
> 
> CI is now passing. Could this PR be merged as-is, with the root fix tracked in #11726?

I don’t think this workaround should be merged. We should fix the real problem instead.

**Comment by JoegenUSTC** [comment]
> > Hi @diqiu50 @yuqi1129, thanks for the valuable feedback!
> > Two updates:
> > 
> > 1. **Null check**: Updated to use `Preconditions.checkArgument(scope != null, "scope cannot be null")` per the Gravitino coding convention, with a dedicated test added to cover the null scope case.
> > 2. **Root fix**: Opened follow-up issue [[Improvement] Ensure all Capability method calls are made within the correct classloader context #11726](https://github.com/apache/gravitino/issues/11726) to track ensuring all `Capability` method calls are made within the correct classloader context — agreed this PR is a workaround and the broader fix deserves its own issue.
> > 
> > CI is now passing. Could this PR be merged as-is, with the root fix tracked in #11726?
> 
> I don’t think this workaround should be merged. We should fix the real problem instead.

Thanks for the clear direction, @diqiu50!

I've implemented the root fix in a new PR: https://github.com/apache/gravitino/pull/11755

The approach wraps the `Capability` delegate in a JDK dynamic proxy inside `CapabilityHelpers.getCapability()`. The proxy's `InvocationHandler` executes every method call inside `classLoader.withClassLoader()`, so the correct classloader context is always active regardless of the `Capability` implementation style — no per-call-site changes needed.

This makes the `switch-on-enum → if-else` change here unnecessary. I'll close this PR once the root fix is accepted. Would appreciate your review on the new PR when you have a moment!

**Comment by yuqi1129** [comment]
@JoegenUSTC 
Could you share which JDK you are using? so we could reproduce the problem you raise.

**Comment by JoegenUSTC** [comment]
> @JoegenUSTC Could you share which JDK you are using? so we could reproduce the problem you raise.

The issue was observed with JDK 17 (the version required by Gravitino's build).

**Comment by yuqi1129** [comment]
> > @JoegenUSTC Could you share which JDK you are using? so we could reproduce the problem you raise.
> 
> The issue was observed with JDK 17 (the version required by Gravitino's build).

More details are preferable, and we have never run into the problem in JDK17(JDK17 is also our debugging JDK version). Can you paste the result of `java --version` here?

**Comment by JoegenUSTC** [comment]
> > > @JoegenUSTC Could you share which JDK you are using? so we could reproduce the problem you raise.
> > 
> > 
> > The issue was observed with JDK 17 (the version required by Gravitino's build).
> 
> More details are preferable, and we have never run into the problem in JDK17(JDK17 is also our debugging JDK version). Can you paste the result of `java --version` here?
@yuqi1129 
/gravitino$ java --version
openjdk 17.0.15 2025-04-15
OpenJDK Runtime Environment Temurin-17.0.15+6 (build 17.0.15+6)
OpenJDK 64-Bit Server VM Temurin-17.0.15+6 (build 17.0.15+6, mixed mode, sharing)

**Comment by JoegenUSTC** [comment]
> > Hi @diqiu50 @yuqi1129, thanks for the valuable feedback!
> > Two updates:
> > 
> > 1. **Null check**: Updated to use `Preconditions.checkArgument(scope != null, "scope cannot be null")` per the Gravitino coding convention, with a dedicated test added to cover the null scope case.
> > 2. **Root fix**: Opened follow-up issue [[Improvement] Ensure all Capability method calls are made within the correct classloader context #11726](https://github.com/apache/gravitino/issues/11726) to track ensuring all `Capability` method calls are made within the correct classloader context — agreed this PR is a workaround and the broader fix deserves its own issue.
> > 
> > CI is now passing. Could this PR be merged as-is, with the root fix tracked in #11726?
> 
> I don’t think this workaround should be merged. We should fix the real problem instead.

@diqiu50 @yuqi1129 — Thanks again for the valuable feedback on both PRs. I wanted to share a new development and get your thoughts.

Following @diqiu50's direction to address the root problem, I opened #11755 with a JDK proxy approach. During review, @yuqi1129 raised a concern — supported by a small standalone test — that the proxy's `withClassLoader` wrapping may not be sufficient for the `$1` case, since the synthetic class is resolved by the defining classloader rather than the TCCL. His test showed the `if-else` version handles this more reliably, even in edge cases where the classloader has already been closed.

This suggests the classloader lifecycle fix (ensuring `IsolatedClassLoader` isn't closed while `Capability` instances are still in use) is a broader problem that deserves its own careful design, now tracked in #11726.

Given this, I'd like to gently revisit whether this PR could be merged as a well-scoped, targeted fix for the immediate problem. I fully respect @diqiu50's preference to avoid landing workarounds, and I remain committed to the lifecycle root fix in #11726. That said, since the root fix will take more time to design and validate properly, would you both be open to landing this as a stable interim improvement while #11726 progresses?

Happy to hear your thoughts — and thank you both for the continued engagement.
